hazelcast/src/main/java/com/hazelcast/map/impl/query/MapQueryEngineImpl.java

```diff
@@ -141,29 +145,42 @@ private Result runQueryOnGivenPartition(Query query, Target target) {
     }
 
     private Result doRunQueryOnQueryThreads(Query query, Collection<Integer> partitionIds, Target target) {
-        Result result = resultProcessorRegistry.get(query.getResultType()).populateResult(query,
-                queryResultSizeLimiter.getNodeResultLimit(partitionIds.size()));
-        dispatchQueryOnQueryThreads(query, target, partitionIds, result);
+        Result result = populateResult(query, partitionIds);
+        List<Future<Result>> futures = dispatchOnQueryThreads(query, target);
+        addResultsOfPredicate(futures, result, partitionIds, false);
         return result;
     }
 
-    private void dispatchQueryOnQueryThreads(Query query, Target target, Collection<Integer> partitionIds, Result result) {
+    private List<Future<Result>> dispatchOnQueryThreads(Query query, Target target) {
         try {
-            List<Future<Result>> futures = queryDispatcher.dispatchFullQueryOnQueryThread(query, target);
-            addResultsOfPredicate(futures, result, partitionIds);
+            return queryDispatcher.dispatchFullQueryOnQueryThread(query, target);
         } catch (Throwable t) {
-            if (t.getCause() instanceof QueryResultSizeExceededException) {
+            if (!(t instanceof HazelcastException)) {
+                // these are programmatic errors that needs to be visible
+                throw rethrow(t);
+            } else if (t.getCause() instanceof QueryResultSizeExceededException) {
```

Comment:
Can `queryDispatcher.dispatchFullQueryOnQueryThread(query, target)` throw an exception with this cause? If I'm not mistaken, that method merely dispatches operations and thus can't receive an exception from the query engine.

Reply:
Some operations are run locally, so within invoke. It's not intuitive, I
agree, but still it holds.